src/term/parser/parser.rs

```diff
@@ -108,31 +111,41 @@ fn token_stream(
 
 // Parsers
 
-fn soft_keyword<'a, I>(keyword: &'a str) -> impl Parser<'a, I, (), extra::Err<Rich<'a, Token>>>
-where
-  I: ValueInput<'a, Token = Token, Span = SimpleSpan>,
-{
-  select!(Token::Name(name) if name == keyword => ())
-}
-
 fn name<'a, I>() -> impl Parser<'a, I, Name, extra::Err<Rich<'a, Token>>>
 where
   I: ValueInput<'a, Token = Token, Span = SimpleSpan>,
 {
-  select!(Token::Name(name) => Name::from(name))
+  // FIXME: bug with chunsky when using with `.repeated`
+  // select!(Token::Name(name) => Name::from(name)).labelled("<Name>")
+
+  any()
+    .filter(|t| matches!(t, Token::Name(_)))
+    .map(|t| {
+      let Token::Name(name) = t else { unreachable!() };
+      Name::from(name)
+    })
+    .or(just(Token::Data).to(Name("data".into())))
```

Comment:
no, i don't think it will produce confusing errors, that error you commented on still happens even when removing the Token::Data and adding back the soft_keyword approach

Reply:
![image](https://github.com/HigherOrderCO/hvm-lang/assets/19491161/31cc686d-f533-456b-83e4-df88c3489ca8)
![image](https://github.com/HigherOrderCO/hvm-lang/assets/19491161/4b06a83e-1f27-4c5e-8aff-80b75fb96061)

Reply:
@developedby here i just changed to use soft-keyword and this are the only changes that happens

Reply:
> ![image](https://private-user-images.githubusercontent.com/19491161/305403673-31cc686d-f533-456b-83e4-df88c3489ca8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDgwODQyNjUsIm5iZiI6MTcwODA4Mzk2NSwicGF0aCI6Ii8xOTQ5MTE2MS8zMDU0MDM2NzMtMzFjYzY4NmQtZjUzMy00NTZiLTgzZTQtZGY4OGMzNDg5Y2E4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjE2VDExNDYwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJkNjY2OTlmYmIzOGM3NTJjYzY3NTRmMWUxZTZlOWNiYzI1MGExN2I0OWYzYjFmZjY1NTRkZjZjYTdjZDg2YzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.wki9BywxYvD5tXaRulDXD1OQaXdGamqPgcaKH23wjXY) ![image](https://private-user-images.githubusercontent.com/19491161/305403751-4b06a83e-1f27-4c5e-8aff-80b75fb96061.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDgwODQyNjUsIm5iZiI6MTcwODA4Mzk2NSwicGF0aCI6Ii8xOTQ5MTE2MS8zMDU0MDM3NTEtNGIwNmE4M2UtMWYyNy00YzVlLThhZmYtODBiNzVmYjk2MDYxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjE2VDExNDYwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU4YWFjNmNlNGUwZjYwMzI2ZjEwMzJhNjkyYjVlMGMxMDEwZGU3ZGIwMjRkZmRlYWNiZDJjYWI1MGE2MDdhODgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.9GoQmrWxdFxk7zEnGnwb8_emPkWGCCSUpaOSt8Zd4yE)

The first message works with the approach you implemented and correctly points out that `*` is not expected.
The second message is wrong in both versions. `(` is valid in this position, it finds EOF but should expect a Name.

Reply:
We can leave it like this then, and create another story for fixing the parser for real, possibly changing to another library